### ISSUE TITLE
BaseChannel/GetBytes, buffer overrun potential

### ISSUE THREAD

**Issue by mr23** [nothingmn/AGENT.Contrib#9]
This needs some improvement to not overrun, and would be more useful if the byte array size was configurable. Also, add BytesToRead to the SP implementation, and use here for performance; push the byte content parsing up. Something like:

public int BufferLen { get; set/\* use a method to change buffer size, protected with a lock, or disallow setting when port is opened*/; }

// don't worry about byte values here, just buffer and return the input, parse in caller.
 protected byte[] GetBytes(SerialPort port)
        {
            int position = 0;
            int bytesWaiting = port.BytesToRead;
            bytesWaiting = Min(bytesWaiting,BufferLen);
            byte[] buffer = new byte[bytesWaiting];
            var bytesWaiting = port.Read(buffer,0,bytesWaiting); // returns bytes read
            return buffer;
        }

**Comment by nothingmn** [comment]
Yes, this stuff is definitely a work in progress!  The whole Channel/Communication setup needs to be ironed out completely and tested fully....

**Comment by mr23** [comment]
I'd also push the buffer out to the class level and preallocate it, for performance. Less GC and frag.

**Comment by mr23** [comment]
Why close the issue if the code is still fragile/incomplete?

**Comment by nothingmn** [comment]
Hit the wrong button.. :p

**Comment by nothingmn** [comment]
This should do it...

```
    protected byte[] GetBytes(SerialPort port)
    {
        byte[] buffer = new byte[port.BytesToRead];
        port.Read(buffer, 0, buffer.Length);
        return buffer;
    }
```

seems simple enough..

what do you think?

**Comment by mr23** [comment]
Could run out of memory quickly. Think 57.6k for example, 1+ seconds. Otherwise it is simply correct. 
In my testing on the emulator, despite either timer polling or DataReceived, sometimes a stall occurred for a few seconds. 
No doubt with an OS the AGENT may too.
Safer to use a limit, allowing User to tweak it.

Sent from my Kindle Fire
